Tests/CachedResponseHandlerTests.swift

```diff
@@ -0,0 +1,226 @@
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import Alamofire
+import Foundation
+import XCTest
+
+final class CachedResponseHandlerTestCase: BaseTestCase {
+
+    // MARK: Properties
+
+    private let urlString = "https://httpbin.org/get"
+
+    // MARK: Tests - Per Request
+
+    func testThatRequestCachedResponseHandlerCanCacheResponse() {
+        // Given
+        let session = self.session()
+
+        var response: DataResponse<Data?>?
+        let expectation = self.expectation(description: "Request should cache response")
+
+        // When
+        let request = session.request(urlString).cacheResponse(using: ResponseCacher.cache).response { resp in
+            response = resp
+            expectation.fulfill()
+        }
+
+        waitForExpectations(timeout: timeout, handler: nil)
+
+        // Then
+        XCTAssertEqual(response?.result.isSuccess, true)
+        XCTAssertTrue(session.cachedResponseExists(for: request))
+    }
+
+    func testThatRequestCachedResponseHandlerCanNotCacheResponse() {
+        // Given
+        let session = self.session()
+
+        var response: DataResponse<Data?>?
+        let expectation = self.expectation(description: "Request should not cache response")
+
+        // When
+        let request = session.request(urlString).cacheResponse(using: ResponseCacher.doNotCache).response { resp in
+            response = resp
+            expectation.fulfill()
+        }
+
+        waitForExpectations(timeout: timeout, handler: nil)
+
+        // Then
+        XCTAssertEqual(response?.result.isSuccess, true)
+        XCTAssertFalse(session.cachedResponseExists(for: request))
+    }
+
+    func testThatRequestCachedResponseHandlerCanModifyCacheResponse() {
+        // Given
+        let session = self.session()
+
+        var response: DataResponse<Data?>?
+        let expectation = self.expectation(description: "Request should cache response")
+
+        // When
+        let cacher = ResponseCacher(
+            behavior: .modify { _, response in
+                return CachedURLResponse(
+                    response: response.response,
+                    data: response.data,
+                    userInfo: ["key": "value"],
+                    storagePolicy: .allowed
+                )
+            }
+        )
+
+        let request = session.request(urlString).cacheResponse(using: cacher).response { resp in
+            response = resp
+            expectation.fulfill()
+        }
+
```

Comment:
I think this PR for Cached Responses looks great.  This will be an excellent addition.  One question I had about the `.modify` case above is that it looks like it allows you to cache a response that was not sent back from the server.  What would be the use case for this?   Would this be fine grained control for the client to reason about the response and then delegate the cache that is saved?

Reply:
Well, that's not really the use case for `.modify`. The `.cache` and `.doNotCache` behaviors are simply convenience cases that pass the cached response or `nil` on to the completion handler automatically. The `.modify` case allows you to do three different things, with full access to all the data that the URL session delegate API provides: 1) cache the response as is, 2) do not cache the response by passing `nil` to the completion, 3) modify the cached response in some way (maybe adjusting the cache-control header) and passing that on to be cached.